cs/di-extensions.texy

```diff
@@ -5,73 +5,115 @@ DI: Rozšíření pro kontejner
 
 /---code neon
 extensions:
-	blog: MyBlogExtension
+	blog: BlogExtension
 \---
 
-Každé rozšíření `Compiler`u musí dědit od [Nette\DI\CompilerExtension|api:] a může implementovat tři různé metody, které jsou postupně volány, během sestavování `Container`u.
+Každé rozšíření `Compiler`u musí dědit od [Nette\DI\CompilerExtension|api:] a může implementovat tři různé metody, které jsou postupně volány, během sestavování `Container`u:
 
+1. loadConfiguration()
+2. beforeCompile()
+3. afterCompile()
 
-CompilerExtension::loadConfiguration() .{toc: loadConfiguration()}
-==================================================================
 
-Metoda je nad všemi rozšířeními volána jako první a je určena k načtení dodatečných konfiguračních souborů, vytvoření dalších služeb pomocí [Nette\DI\ContainerBuilder|api:] a hlavně zpracování konfigurace aplikace.
+loadConfiguration() .[method]
+=============================
 
-V configu je možné uvést sekci, která se jmenuje stejně jako naše rozšíření. Pokud tedy budeme uvažovat předchozí příklad, mohlo by v konfigu přibýt například toto
+Tato metoda se volá jako první a je určena k přidání služeb souvisejících s daným rozšířením.
+
+Na registrování služeb použijeme [Nette\DI\ContainerBuilder|api:].
+
+/--php
+class BlogExtension extends Nette\DI\CompilerExtension
+{
+	public function loadConfiguration()
+	{
+		$builder->addDefinition($this->prefix('articles'))
+			->setFactory('App\Model\HomepageArticles', ['@connection'])
+			->addSetup('setLogger', ['@logger']);
+	}
+}
+\--
+
+.[tip]
+Konvence je, prefixovat služby v rozšíření jeho názvem, aby nevznikaly konflikty. Pomůže nám s tím metoda `prefix()`.
+
+
+Konfigurace rozšíření
+---------------------
+
+Rozšíření můžeme konfigurovat v sekci, jejíž název je stejný jako ten, pod kterým bylo rozšíření přidáno, tedy *blog*.
 
 /--neon
-blog: # stejné jméno jako má extension
+# stejné jméno jako má extension
+blog:
 	postsPerPage: 10
 	comments: false
 \--
 
-Nastavení si lze přečíst v proměnné `$this->config` v rozšíření.
+K definovaným hodnotám se v rozšíření dostaneme přes proměnnou `$this->config`.
 
 /--php
-class MyBlogExtension extends Nette\DI\CompilerExtension
+class BlogExtension extends Nette\DI\CompilerExtension
 {
 	public function loadConfiguration()
 	{
-		dump($this->config);
+		$config = $this->config;
 		// [2] [ 'postsPerPage' => 10, 'comments' => false ]
+	}
+}
 \--
 
-Protože uznáváme princip [konvence před konfigurací |http://en.wikipedia.org/wiki/Convention_over_configuration], nastavíme si výchozí hodnoty, aby aplikace fungovala i bez toho, že budeme něco nastavovat, a sloučíme je hodnotami z konfiguračního souboru.
+
+Výchozí hodnoty
+---------------
+
+Protože uznáváme princip [konvence před konfigurací |http://en.wikipedia.org/wiki/Convention_over_configuration], 
```

Comment:
Pokud to lze, je užitečné podle principu [konvence před konfigurací |http://en.wikipedia.org/wiki/Convention_over_configuration]
jednotlivým prvkům konfigurace poskytnout výchozí hodnoty, aby uživatel nemusel konfigurovat úplně vše.
Pro sloučení uživatelské konfigurace a výchozích hodnot můžeme použít metodu `validateConfig()`. Jako parametr jí předáme pole všech povolených klíčů a jejich výchozích hodnot, o které rozšíří `$this->config`. Zároveň pokud se v konfiguraci objeví klíče, které nejsou mezi povolenými, bude vyhozena výjimka. Tak se předejde tomu, že někdo omylem splete název klíče v konfiguračním souboru.